### PR TITLE
Refactoring ckb4ibc-test with ibc test framework

### LINKED ISSUE #263: upgrade CI ckb4ibc-test to adapt relay feature between Axon and Ckb
## Summary
after successfully complete connection and channel layer between Axon and Ckb, we should enable corresponded CI process to automatically check its correctness in every PR

## Proposal

Review and Refactor https://github.com/synapseweb3/forcerelay/tree/dev/tools/integration-test 

## References
- IBC Relayer Integration Test Framework 
    https://github.com/informalsystems/hermes/blob/master/tools/test-framework/README.md

### PR DESCRIPTION
**Pull request by jjyr**
## Description

This PR tries to introduce [ibc-test-framework](https://github.com/informalsystems/hermes/blob/master/tools/test-framework/README.md) to `ckb4ibc-test`.

This PR is focus on move the exists testcase into ibc test framework.

Most code is copied from https://github.com/synapseweb3/forcerelay/blob/d1057ee7c8cdd3d96de146396e4225f5dd8f5d24/tools/ckb4ibc-test/src/lib.rs without modification.

The code qualification is low, I will send another PR to focus on the code qualification improvement. I believe split into two PRs can help reviewers.

### Known limitation

The `ibc-test-framework` has its essential limitations, it assumes the chain is a `gaiad` liked chain. To make it work with CKB or Axon we need to modify some code, mostly modifications are in the [tools/ckb4ibc-test/src/framework](https://github.com/synapseweb3/forcerelay/compare/dev...refactor-ckb-4ibc-test-use-ibc-test-framework?expand=1#diff-62a847c3216e2d3fe81fff9f12976431cc927bad87406873d33063507a78440d) folder.

This PR focus implements the basic testcase - `create channels`, however we still not fully integrate with the framework since we have hardcoded parts:

1. the private key of chain relayer is hardcoded
2. the contract deployment transaction is hardcoded

We need further PRs to improve the contract deployment feature.

### Known bugs:

In the original testing code the node A open connection and channel to node B. the two nodes never try to initialize a connection/channel in the same time, but this isn't the behavior in the current testcase.
In current testcase the node A and node B tries to initialize connection / channel at the same time. This help us found two bugs:

1. when the node B receives `InitConnection` from A, the `InitConnection` initiated from B is failed, but the incompleted connection is never deleted from the `connections` vector.
2. when the node B receives `InitChannel` from A, the `InitChannel` initiated from B is failed, its state is stucked on the `Init` process.

We need further PRs to dig into the two bugs.

### CODE REVIEW

**Review by chaoticlonghair — APPROVED** [review]

**Comment by ashuralyk** [comment]
> the private key of chain relayer is hardcoded

what does this mean, we read private key from config file and it's changeable I think

> the contract deployment transaction is hardcoded

the solidity and ckb contracts are assumed to be developed before the initialization of Forcerelay, so this is the `hardcoded` meaning here ?

> Known bugs

there known bugs are triggered in solidity contract of ckb contract ?

File: tools/ckb4ibc-test/src/framework/utils/ckb.rs [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+        return false;
+    }
+    // FIXME @jjy in current implementation,
+    // a failed init conn is remained in the vector, so we must check last opened conn
+    let connection = connection.connections.into_iter().last().unwrap();
+    connection.state == State::Open
+}
+
+fn send_tx(req: &str, port: u32) -> RpcResponse {
+    let output = Command::new("curl")
```
Comment (blckngm) [inline tools/ckb4ibc-test/src/framework/utils/ckb.rs:60]: Why not use reqwest?

Comment (jjyr) [reply tools/ckb4ibc-test/src/framework/utils/ckb.rs:60]: Copied from the original code, we should do the refactoring in later PRs.

File: tools/ckb4ibc-test/src/framework/utils/ckb.rs [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+            let mut cell = issued_cell_template.clone();
+            cell.lock.args = JsonBytes::from_vec(args.to_vec());
+            dev_spec.genesis.issued_cells.push(cell);
+        }
+        let content = toml::to_string_pretty(&dev_spec).unwrap();
+        fs::write(dev_spec_path, content).unwrap();
+    }
+
+    if port != 8114 {
+        Command::new("sed")
```
Comment (blckngm) [inline tools/ckb4ibc-test/src/framework/utils/ckb.rs:129]: Is it possible to use `toml_edit` instead?

Comment (jjyr) [reply tools/ckb4ibc-test/src/framework/utils/ckb.rs:129]: Agreed, but this is also copied from the old code, I want to fix it in later PRs.

File: tools/ckb4ibc-test/src/framework/utils/ckb.rs (hunk 1 above)
Comment (blckngm) [reply tools/ckb4ibc-test/src/framework/utils/ckb.rs:60]: OK.

**Comment by jjyr** [comment]
> we read private key from config file and it's changeable I think

The deployment transaction is hardcoded and signed with the specific pk, so we can't simply change the key without change the deployment transaction.

https://github.com/synapseweb3/forcerelay/tree/main/tools/ckb4ibc-test/txs

we need to be able to sign the deployment transaction with arbitrary key, do we support deploy via cli? I haven't digging into it.

> there known bugs are triggered in solidity contract of ckb contract ?

Not the contract, The bug is related to how `ckb4ibc` handle it's local state when receives connection/channel. You can uncommend the code https://github.com/synapseweb3/forcerelay/blob/be6c2fbbd5ced5e883cbeeee7f0e1331af913901/tools/ckb4ibc-test/src/tests/channel.rs#L55

and see the test become failed.

hint: you can use `dbg!(&connection.connections)` to see the connections or channel state of node B. @ashuralyk

**Comment by jjyr** [comment]
I spent some time digging into the hermes code.

I guess the known-issue-1 is caused by [this function](https://github.com/synapseweb3/forcerelay/blob/dev/tools/test-framework/src/bootstrap/binary/connection.rs#L40-L53), here is the analysis:

1. The purpose here is to create dummy connections, so the connection id is not always `0` in the test. (original comment form hermes, https://github.com/synapseweb3/forcerelay/blob/dev/tools/test-framework/src/bootstrap/binary/connection.rs#L91-L99)
2. In in this method, the `pad_connection_id_b` is set to `1`. https://github.com/synapseweb3/forcerelay/blob/dev/tools/test-framework/src/bootstrap/binary/connection.rs#L141 
3. so in the log we observe the chainB is always start with an InitConnection event. Which is the reason why chainB has an unused connection.

So this behavior  is intentioned by the ibc-test-framwork, the purpose is to catch bugs, I will send a PR to update the comment.

**Comment by ashuralyk** [comment]
the `PRIVKEY` in ckb4ibc-test is fixed, because the related ckb account is written into genesis block of dev-chain, so this cannot be changed until we set a process to change the config file of this dev-chain

**Pull request merged**